### PR TITLE
refactor away from global context states

### PR DESCRIPTION
**Pull request by xg-wang**
Avoid relying on mutating global states, create beacon and its associated database objects instead.

### CODE REVIEW

File: src/queue.ts [hunk 1]
```diff
@@ -81,7 +74,16 @@ export class QueueImpl implements Queue {
       .then((entries) => {
         if (entries.length > 0) {
           const { url, body, timestamp, statusCode, attemptCount } = entries[0];
-          return fetchFn(url, body, createHeaders(attemptCount, statusCode))
+          debug(
```
Comment (asakusuma) [inline src/queue.ts:77]: how are these debug statements stripped out for prod build?

File: src/queue.ts [hunk 2]
```diff
@@ -165,50 +167,45 @@ class NoopQueue {
...
+ * @public
  */
-export function onClear(cb: () => void): void {
-  beaconListeners.add(cb);
-}
-export function removeOnClear(cb: () => void): void {
-  beaconListeners.delete(cb);
-}
+export class RetryDB {
+  static hasSupport =
```
Comment (asakusuma) [inline src/queue.ts:174]: oh this is a nice technique

File: src/beacon.ts [hunk 3]
```diff
@@ -129,23 +137,40 @@ class Beacon {
...
 
-export default beacon;
+function prepareConfig(init: BeaconInit): Required<BeaconInit> {
+  const beaconConfig = init.beaconConfig || {
+    retry: {
+      limit: 0,
+    },
+  };
+  const retryDBConfig = init.retryDBConfig || {
+    storeName: 'default',
```
Comment (asakusuma) [inline src/beacon.ts:165]: I would use a different default name since isn't this a global name for everything on the domain?

**Review by asakusuma — CHANGES_REQUESTED** [review]
thanks

File: src/queue.ts (hunk 1 above)
Comment (xg-wang) [reply src/queue.ts:77]: The debug statements aren't extracted, there's a debug flag to enable logging in prod build

File: src/beacon.ts (hunk 3 above)
Comment (xg-wang) [reply src/beacon.ts]: Actually objectStore name should not be customized by consumer, database name is. I changed in the latest commit.

**Review by asakusuma — APPROVED** [review]
thanks!

**Pull request merged**